### PR TITLE
2D FoV: Fix a bug and represent slopes with integers.

### PR DESCRIPTION
**Pull request by KyleSiefring**
Fixes a bug that caused casted light to veer off in some cases.
Improve precision by replacing floats with pairs of integers when representing slopes.

### CODE REVIEW

File: src/lightmap.cpp [hunk 1]
```diff
@@ -492,6 +492,46 @@ bool map::pl_line_of_sight( const tripoint &t, const int max_range ) const
...
+}
+
+inline bool operator< ( const slope &lhs, const slope &rhs )
+{
+    // x1/y1 < x2/y2 === x1*y2 < x2*y1 if y1 and y2 have the same sign.
+    return lhs.xdiff * rhs.ydiff < rhs.xdiff * lhs.ydiff;
+}
+inline bool operator> ( const slope &lhs, const slope &rhs )
+{
+    return rhs < lhs;
```
Comment (codemime) [inline src/lightmap.cpp]: Is there any reason not to write `lhs > rhs` (which is easier to comprehend)?

File: src/lightmap.cpp [hunk 2]
```diff
@@ -878,9 +918,12 @@ void castLight( float (&output_cache)[MAPSIZE*SEEX][MAPSIZE*SEEY],
         for( delta.x = -distance; delta.x <= 0; delta.x++ ) {
             int currentX = offsetX + delta.x * xx + delta.y * xy;
             int currentY = offsetY + delta.x * yx + delta.y * yy;
-            float trailingEdge = (delta.x - 0.5f) / (delta.y + 0.5f);
-            float leadingEdge = (delta.x + 0.5f) / (delta.y - 0.5f);
+            slope trailingEdge( delta.x * 2 - 1, delta.y * 2 + 1 );
+            slope leadingEdge( delta.x * 2 + 1, delta.y * 2 - 1 );
```
Comment (codemime) [inline src/lightmap.cpp]: Might be a premature optimisation, but I'd write `delta.x shl 1` here.

File: src/lightmap.cpp (hunk 1 above)
Comment (narc0tiq) [inline src/lightmap.cpp]: Uh... because that would just make it call itself until it fills up the call stack?

**Comment by KyleSiefring** [comment]
Okay, world made that demonstrates the bug. In master there is a sliver of vision that shouldn't exist.
[Marlinton.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/275998/Marlinton.zip)

File: src/lightmap.cpp (hunk 2 above)
Comment (KyleSiefring) [inline src/lightmap.cpp]: The compiler should optimize do this anyway.

File: src/lightmap.cpp (hunk 1 above)
Comment (codemime) [inline src/lightmap.cpp]: Ha-ha, right. That was the silliest question ever.

File: src/lightmap.cpp [hunk 3]
```diff
@@ -492,6 +492,46 @@ bool map::pl_line_of_sight( const tripoint &t, const int max_range ) const
 
 #include "messages.h"
 
+typedef ShadowcastingSlope slope;
+
+slope::ShadowcastingSlope( int x, int y )
```
Comment (BevapDin) [inline src/lightmap.cpp]: That looks like a constructor (and actually is one), but it has a strange name. Why are you using different type names here? It should be either `slope::slope` or `ShadowcastingSlope::ShadowcastingSlope`.

Comment (masecknotrealname) [inline src/lightmap.cpp]: I tried doing slope::slope but it didn't work. ShadowcastingSlope::ShadowcastingSlope is probably better in this case.

Comment (kevingranade) [inline src/lightmap.cpp]: Why are you even doing a typecast? If the name is too long make it not so
long, otherwise just use it as-is.

Comment (KyleSiefring) [inline src/lightmap.cpp]: I removed the typedef.

**Comment by kevingranade** [comment]
I can't download the save, and I'm not sure I'd know what to look for in any case.  Can you describe the line, like the relative coordinates of the player and any obstacles needed to create the situation?
Ideally I'd like to add this to the unit tests.
Also @Coolthulhu this may well be the cause of the artifacts you noticed?

**Comment by kevingranade** [comment]
I'm wondering if merely increasing the resolution of the number to a double will fix the issue as well.

**Comment by KyleSiefring** [comment]
Upon further examination, it seems that using floats only causes problems if you calculate the reciprocal separately. I was doing that at some point. Using a double for the reciprocal does seem to fix the problem though. Optimizing the integer version would probably give faster code, but this patch is meant to fix bugs.

I verified it with this code.
```
    constexpr int row_size = 128;
    constexpr int size = row_size * row_size;
    float float_edges[size];
    ShadowcastingSlope int_edges[size];
    for( int y = 0; y < row_size; y++ ) {
        for( int x = 0; x < row_size; x++ ) {
            float_edges[y * row_size + x] = (x + .5) / (y + .5);
            int_edges[y * row_size + x] = ShadowcastingSlope( 2*x + 1, 2*y + 1 );
        }
    }
    for ( int i = 0; i < size; i++ ) {
        for ( int j = 0; j < size; j++ ) {
            bool less = float_edges[i] < float_edges[j] ^ int_edges[i] < int_edges[j];
            bool equal = float_edges[i] == float_edges[j] ^ int_edges[i] == int_edges[j];
            bool greater = float_edges[i] > float_edges[j] ^ int_edges[i] > int_edges[j];
            if (less || equal || greater) {
                printf("%d %d\n", i, j);
            }
        }
    }
```

Here is an image of the bug that this patch does fix.
startlessend

Relevant code.
```
+                // Ensure the new start slope is still valid.
+                if( start < end ) {
+                    return;
+                }
```

**Comment by kevingranade** [comment]
I recreated your scenario and I'm not seeing that artifact. Was that with
the upstream code or after some modifications?
I'm seeing some other issues though so I'm still looking into it.

**Comment by KyleSiefring** [comment]
That was on master. Are you sure you recreated it right? If you turned on 3d fov that would also mess it up. The 3d code iterates in the opposite direction (i.e. horizontal/vertical to diagonal).

**Comment by kevingranade** [comment]
I was running with 3D FoV, and it was creating other artifacts unfortunately.
I'm putting together a little framework for unit testing the shadowcasting implementation.
Preliminary results show that your start < end check does fix the problem you outline,. but I don't see a rationale for the float -> pair conversion.

**Comment by mugling** [comment]
Closed as appears stalled. Needs updates and rebase if OP wants to resume

**Pull request closed without merging**